packages/mdc-textfield/index.ts

```diff
@@ -0,0 +1,451 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import {MDCComponent} from '@material/base/component';
+import {EventType, SpecificEventListener} from '@material/base/types';
+import * as ponyfill from '@material/dom/ponyfill';
+import {MDCFloatingLabel} from '@material/floating-label/index';
+import {MDCLineRipple} from '@material/line-ripple/index';
+import {MDCNotchedOutline} from '@material/notched-outline/index';
+import {MDCRippleFoundation} from '@material/ripple/foundation';
+import {MDCRipple} from '@material/ripple/index';
+import {RippleCapableSurface} from '@material/ripple/types';
+import {MDCTextFieldCharacterCounter, MDCTextFieldCharacterCounterFoundation} from './character-counter';
+import {cssClasses, strings} from './constants';
+import {MDCTextFieldFoundation} from './foundation';
+import {MDCTextFieldHelperText, MDCTextFieldHelperTextFoundation} from './helper-text';
+import {MDCTextFieldIcon} from './icon';
+import {
+  CharacterCounterFactory,
+  FoundationMapType,
+  HelperTextFactory,
+  IconFactory,
+  LabelFactory,
+  LineRippleFactory,
+  OutlineFactory,
+  RippleFactory,
+} from './types';
+
+class MDCTextField extends MDCComponent<MDCTextFieldFoundation> implements RippleCapableSurface {
+  static attachTo(root: Element): MDCTextField {
+    return new MDCTextField(root);
+  }
+
+  // Public visibility for these properties is required by RippleCapableSurface.
+  root_!: HTMLElement; // assigned in MDCComponent constructor
+  ripple!: MDCRipple | null; // assigned in initialize()
+
+  // The only required sub-element.
+  private input_!: HTMLInputElement; // assigned in initialize()
+
+  // Optional sub-elements.
+  private characterCounter_!: MDCTextFieldCharacterCounter | null; // assigned in initialize()
```

Comment:
Using `null` just made the code a little cleaner by condensing the `if` statements down to ternaries. That's the only reason I did it 😄

Reply:
I guess since they're all private it doesn't really matter either way. I'm not sure if there's any perception that one is preferable than the other, typing-wise, other than `?` existing as a more concise representation of `| undefined`.